### PR TITLE
[proof] Rework `proof_mode` handling.

### PR DESCRIPTION
**Pull request by ejgallego**
**update: this PR has changed significantly, see: https://github.com/coq/coq/pull/459#issuecomment-293729031**

This patch reworks "proof_mode" handling, moving the state itself to
the parsing summary, and giving the decision making to the STM.

When parsing, the STM will try to classify statically the start of a
proof. If the command can be statically classified, the STM will set
the parsing proof mode itself and allow us to continue parsing.

If the STM cannot classify the proof mode, it will indeed run the
command, and `Lemmas.start_proof/save_proof` will be the ones
doing the push/pop of proof mode states.

### CODE REVIEW

**Comment by ejgallego** [comment]
So I have become definitively convinced that in order to support the short form of the proof modes CEP, we should move the classification of proof-starting commands to the parser.

What do you think @gares ?

Unfortunately it seems we cannot classify `VernacInstance` and some parts of obligations as `VtStartProof`, that would be great as we could move the classification to the parser just by assuming that VtStartProof will set the proof mode, and build upon a refinement of that approach.

However we may need to explicity add a `starts_proof : vernac_expr -> bool` classifier just to be more reliable and not depend on the hack in `Lemmas.start_proof`.

The good thing is that later on, we could evolve to `starts_proof : vernac_expr -> proof_mode` and voilà.

**Comment by ppedrot** [comment]
I am not an expert in the STM, so I can't really tell what's happening, but I think you're throwing the baby with the water in this patch. If I understand correctly, one of the issues you're facing is that the proof mode is stored in an ephemeron, so that the STM loses it sometimes and goes awry. But with your patch, the current mode is not stored anywhere, as the only thing you keep in the summary is a bool. Why don't you rather keep a string option which stands for the currently enabled mode (if any) and register the grammar in some global, non-synchronized table? In particular, your patch removes the option to change the proof mode on the fly...

**Comment by ejgallego** [comment]
Hi Pierre-Marie, here I disagree a bit, I'll try to explain my point of view:

- There are two things here, the old "proof_mode", and the new desired "tactic parsing mode". IMHO they work at a fundamentally different level. In my opinion it makes sense to remove the old implementation, first because it was based on the `unit -> unit` API which is a bit annoying, and secondly, because it was fundamentally broken when put in conjunction with the STM model. The old implementation relied in `start_proof` and friends, by in an asyncronous execution model that doesn't happen when you think. Indeed, if you trace the `proof_mode` event, you will see what I mean. For a single command things may be set and unset several times, as `proof` and the STM fight to have something reasonable.
- Moreover, the old clasiffication scheme was a bit tricky wrt to proof mode, in the sense that `VtUnknown` commands bascially told nothing about what the command should do on the proof mode, leaving setting it to the vernac execution itself.
- Also, the model precluded having things such as tactic namespaces `(intros; p)%ltac2`.

So we have a model where 3 components try to fight to control the proof mode. Now the real question is, "who is in charge of the proof mode"? Given that it is mostly a parsing issue, and that we'd like to integrate with the notation namespaces, I think it makes sense to let the parser choose.

Deciding this in indeed the crux of this RFC. Do we all think it makes sense to have a model where the parser can statically know which commands will start a proof and what such proof system will be?

After a couple of weeks thinking on this (I got drawn into this due to the fact the `Stm.observe` is buggy wrt the proof mode ! ), I think indeed this is the simpler and more powerful approach. You put the tactic parsing mode in the parser and store it in its state.

Now, about the status of the current patch:
- It is buggy, we are recording the mode in the parser and indeed, due to the interaction with `start_proof` this fails as the mode set may be lost before freezing. AFAICS the only reliable solution is to move the classification of commands starting proofs to the parser even now. I'll try to complete this as soon as I have a minute.
- The point about the mode being a boolean it is a transient one as noted in the comments. We would make that a string or whatever is needed once the new API for declaring tactic parsing modes would be defined. Indeed once the command classification issue is solved we should re-add proof mode and proof mode switching in the style of the CEP.
- Epheremons have nothing to do, the STM stores the proof mode on its on in proof nodes using a custom field. Indeed, interactions with `start_proof` make difficult to put the mode in the Summary.

Thanks for the feedback let me know what you think

**Comment by ejgallego** [comment]
Ok, so I've pushed a revised version of this patch, which is indeed now much more robust, and fixes IMVHO in a proper way longstanding `Load` problems.

If you look at `pcoq.ml`, the main remaining point is indeed definining the proper `VERNAC COMMAND EXTEND` and friends syntax so we can specify a tactic table and whether the command starts a proof.

I wonder if  `grammar_extend` should be deprecated, or would imply too many problems with plugins.
For sure, it looks like `extend_grammar_command` should be modified to take the table parameter.

Let me know what you think folks, IMVHO, once we sort out the proper `VERNAC  EXTEND` and friends syntax, having this working is close.

**Comment by ejgallego** [comment]
Ok we have modified this PR significantly to be more conservative. The code runs but is still not fully finished as it is pending some changes in the PR from @gares .

File: proofs/proof_global.ml [hunk 1]
```diff
@@ -247,10 +169,10 @@ let start_proof sigma id ?pl str goals terminator =
     endline_tactic = None;
     section_vars = None;
     strength = str;
-    mode = find_proof_mode "No";
     universe_binders = pl } in
   push initial_state pstates
 
+(* This  *)
 let start_dependent_proof id ?pl str goals terminator =
```
Comment (ppedrot) [inline proofs/proof_global.ml]: Truncated comment maybe?

Comment (ejgallego) [inline proofs/proof_global.ml]: Indeed @ppedrot , you may want to have a look to the only user of this function and in particular at `Proof.dependent_start`, etc..., we are not sure this and their associated machinery should be kept in place.

Comment (ejgallego) [inline proofs/proof_global.ml]: Comment updated thanks.

File: parsing/pcoq.ml [hunk 2]
```diff
@@ -587,12 +587,31 @@ let recover_grammar_command (type a) (tag : a grammar_command) : a list =
 
 let extend_dyn_grammar (GrammarCommand.Dyn (tag, g)) = extend_grammar_command tag g
 
+(* Tactic parsing modes *)
+let add_proof_tactic_entry p e = Hashtbl.add tactic_entry p e
```
Comment (ppedrot) [inline parsing/pcoq.ml]: You should probably add an assert not to register two entries with the same name.

Comment (ejgallego) [inline parsing/pcoq.ml]: Indeed, the API is still very preliminar so I expect it'll have to greatly evolve to be able to put GEXTEND on their own table.

A harder task will be to switch parser state using a rule...

Comment (ppedrot) [inline parsing/pcoq.ml]: That was going to be my question: so you want to switch proof mode while parsing a sentence? I understand the motivations, but *timeo CamlpX et dona ferentem*. If there is a part of Coq I deeply mistrust, this is parsing indeed. Probably after unification, modules, universes, and a few more components. But still, I'm concerned by the fragility and the scalability of such a feature.

**Comment by ejgallego** [comment]
> you want to switch proof mode while parsing a sentence?

For now they will have to be switched using a vernacular. Roughtly, I think a roadmap could be:
- Complete this PR which cleans up the tactic/non-tactic mode handling and fixes a couple of bugs in the STM (sadly not the `Load` problems yet, but maybe we get there).
- Provide experimental support for switching parsing tables, both for notations and proofs. This will be done throu vernaculars. I am particularly interested in switching the parser for different notations. IMHO it is important to have this extremely experimental feature ready so some users can test it and give feedback.

I think that this feature will take a long time to be ready, and I agree with your views on parsing. In particular we need to consider the overlap with the functionality of the module system (but we'd need unimport).

Regarding parsing switching _on-the-fly_, indeed currently it can only be done either with a manual parser implementation or by hacking the error recovery mechanism, but I'd say that it is likely that for when the time we have gathered enough feedback from interested people, parsing has switched to PPX and maybe a mode switch could be added to rules.

File: parsing/pcoq.ml (hunk 2 above)
Comment (ejgallego) [inline parsing/pcoq.ml]: Actually despite my comment below, I think there is a way to do that without incurring into awful sins, but I need to try.

File: stm/stm.ml [hunk 3]
```diff
@@ -496,21 +494,14 @@ end = struct (* {{{ *)
...
+      checkout edit_branch
     end else
       let pl = proof_nesting () in
       try
-        let branch, mode = match Vcs_aux.find_proof_at_depth !vcs pl with
-          | h, { Vcs_.kind = `Proof (m, _) } -> h, m | _ -> assert false in
-        checkout branch;
-        stm_prerr_endline (fun () -> "mode:" ^ mode);
-        Proof_global.activate_proof_mode mode
+        let branch = fst @@ Vcs_aux.find_proof_at_depth !vcs pl in
```
Comment (gares) [inline stm/stm.ml]: We have the $ in caml?  Since when?!

**Review by gares — APPROVED** [review]

File: stm/stm.ml (hunk 3 above)
Comment (Zimmi48) [inline stm/stm.ml]: We have it but I can't stand this syntax. Can you use instead:

    let branch = Vcs_aux.find_proof_at_depth !vcs pl |> fst in

? This is the style that Matej has been introducing bit by bit and I support him on that.

Comment (matejkosik) [inline stm/stm.ml]: I think this is a kind of little-endian / big-endian problem in Ocaml context.

There is a group of people who avoids these operators entirely.

Then there are people who prefer `@@`.

There is a group of people who [oppose](https://github.com/ocaml/ocaml/pull/791) to usage of, GASP, functional composition.

And then there are people who prefer `|>` and `%>`.
(I belong to this group. I like to see the operations in  the same as pipelines in UNIX).

**Comment by gares** [comment]
#566 should supersede this one IMO.

File: stm/stm.ml (hunk 3 above)
Comment (gares) [inline stm/stm.ml]: $, or @@, is nice IMO. It is like opening `(` and saying "until the end of the line".
It was invented quite some time ago in Haskell.
I quite don't get why `@@` was preferred to `$`, but still..

`|>` has the nice property of putting the subject upfront, like the `|` of unix.

IMO, depending on the situation and the stack depth one or the other may work better.
Here, for example, the stack is 1 slot and contains `fst`, so I can stand it ;-)

**Comment by silene** [comment]
There is something I don't understand about this pull request. From a user point of view, it was kind of making sense to be able to type `Proof Mode Foo` to switch the mode to Foo. You removed this feature and did not replace it with anything else, as far as I can tell. So it seems I am missing something here. For instance, how is the user supposed to pass from the classic mode to a separation logic mode?

**Comment by gares** [comment]
This PR (and #566 consequently) remove the concrete syntax to switching mode, or better it can be switched only by starting a proof now (not in the middle of it).

My guess is that another PR about proof_modes should introduce a more compact one for switching proof mode.

**Comment by gares** [comment]
I re-checked the code an indeed the removal of ProofMode was a bit too drastic.
I'll try to update #566 accordingly.

**Comment by ejgallego** [comment]
I had the same doubts, but `VernacProofMode` is currently both broken and dead code, and doesn't work at all for the intended used cases. I suggest to remove it.

**Comment by gares** [comment]
I've updated #566 so that `VernacProofMode` is back and, thanks to the approach proposed by this PR and refined in #566, it works™

I've added a test for CoqIDE and for coqtop (a bit tricky to get it integrated in our build system, also the ML API to inhabit a non-terminal is lacking).  It is based on a plugin that registers a proof mode containing only the "test." tactic and the scripts switch back and forth, also nesting proofs.

**Comment by Zimmi48** [comment]
Should this one be closed in favor of #566? (Even if the latter is not ready.)

**Comment by ejgallego** [comment]
> Should this one be closed in favor of #566? (Even if the latter is not ready.)

Nope, sorry for the lack of updates on this. We need to work more on this and there is still some possibility that this branch gets to a mergeable state.

**Comment by maximedenes** [comment]
@ejgallego Should this PR be closed? I see you moved it to "obsolete".

**Comment by ejgallego** [comment]
Obsolete doesn't mean "to close"; its meaning is more "we tried and couldn't make it work with the current approach yet"; so this is blocking on other changes that may indeed leave this PR truly obsolete, and then be closed; however, it shouldn't be closed until the changes happen.

**Comment by ejgallego** [comment]
For now I am planning to finish #6171 before revisiting this PR, so stuck on it.

**Comment by ejgallego** [comment]
I'm much afraid this PR is not going to work in the current shape, a full rewrite is needed. #566 could have a better chance to work I think.

**Pull request closed without merging**